File: observer/cpp/observer_test.cc
#include <gtest/gtest.h>
#include "observer/cpp/observer.h"
#include <vector>

TEST(ObserverTest, ReaderTest) {
    Subject subject = Subject();
    // 这里不能用Reader然后emplace_back(&readers[i])，vector动态扩容的时候地址会变！
    std::vector<Reader*> readers;
    for(int i=0; i<3; ++i) {
        readers.emplace_back(new Reader());
        subject.attach(readers[i]);
    }
    subject.update_context(std::string("hello reader"));
    for(int i=0; i<3; ++i) {
        EXPECT_EQ(readers[i]->get_message(), "hello reader");
        delete readers[i];
    }
}

File: observer/observer_test.go
package observer

import "testing"

func TestObserver(t *testing.T) {
	subject := NewSubject()
	reader1 := NewReader("reader1")
	reader2 := NewReader("reader2")
	reader3 := NewReader("reader3")
	subject.Attach(reader1)
	subject.Attach(reader2)
	subject.Attach(reader3)

	if reader1.GetMessage() == "observer mode" {
		t.Fail()
	}

	if reader2.GetMessage() == "observer mode" {
		t.Fail()
	}

	if reader3.GetMessage() == "observer mode" {
		t.Fail()
	}

	subject.UpdateContext("observer mode")

	if reader1.GetMessage() != "observer mode" {
		t.Fail()
	}

	if reader2.GetMessage() != "observer mode" {
		t.Fail()
	}

	if reader3.GetMessage() != "observer mode" {
		t.Fail()
	}
}
